### ISSUE TITLE
Implement parser for detecting Autometrics annotations

### ISSUE THREAD

**Issue by arendjr** [autometrics-dev/vscode-autometrics#29]
Currently the Python implementation uses a regular expression to detect Autometrics annotations. In order to implement Rust support, this approach would not be sufficient, since Rust can annotate entire `impl` blocks with a single annotation, which would be too complex to parse with a regex.

To solve this, we need to implement a proper parsing solution in the extension. Something like [Tree-sitter](https://tree-sitter.github.io/tree-sitter/) might be an option, since it supports all the languages we want to support as well, but other options may be fine too.

**Comment by emschwartz** [comment]
VS Code's docs provide an example of having one language server extension hijack requests of a certain type and forward some of those requests to the underlying language server. It seems like we should be able to use this approach to have an autometrics language server that intercepts requests, forwards them as necessary, and injects additional stuff into the response when it's appropriate https://code.visualstudio.com/api/language-extensions/embedded-languages#request-forwarding

**Comment by emschwartz** [comment]
Also, I really think we should try to use the LSP for as much of this as we can. The Language Server Protocol was designed to allow additional information to be added into the editor in a cross-editor way. If we can figure out how to leverage it, hack it, or stretch the bounds of what it was meant for, it seems like we'd be able to build something that would work across editors. This is important, as we've already gotten a [request for Vim support](https://github.com/orgs/autometrics-dev/discussions/31) and other people will want support for editors.

**Comment by flenter** [comment]
I agree that an LSP approach might be the long term solution for several languages. Though not for all languages, for instance with typescript we can communicate with typescript server and get information about the current type through there. We should evaluate for each language which approach would work best.
